proposals/0027/README.md

```diff
@@ -0,0 +1,150 @@
+
+
+## System-context-diagram
+
+_(How does this feature or discussion fit into the big picture? Tinkerbell, like
+many other modern software stacks, have a high level of interaction between
+component, including external components. A system context diagram helps the
+reader visually understand how this Proposal plays a role.)_
+
+As a use-case, we intend to create custom Tinkerbell actions after updating Hegel. It will require some effort to make the Hegel metadata available, but once this is done, one very useful and simple action would query all Hegel metadata and render it into a file that could be indexed. This could be used to write any arbitrary template in Tinkerbell. 
+
+*Example:*
+```yaml
+image: WriteTemplatesFile
+params:
+    DISK: "{{ .disks[0] }}"
+    PATH: /etc/network/01-tinkerbell.cfg
+    CONTENTS: |
+        network:
+            eno1:
+                ip: {{ ."1a:1a:1a:1a:1a".ipv4[0].ip }}
+                netmask:  {{ ."1a:1a:1a:1a:1a".ipv4[0].netmask }}
+```
+
+The following is a diagram depicting how a Tinkerbell will leverage Hegel related to the use-case above. 
+
+![tinkerbell action](tink-action-workflow.png)
+
+## APIs
+
+The following table details the endpoints of instance metadata. Some of the endpoints include placeholders for data that is unique to your instance. 
+
+|Endpoint|Description|Version|
+|--------|-----------|-------|
+|/v0/meta-data/||	|
+|/v0/meta-data/hostname|The hostname of the instance.|	v0|
+|/v0/meta-data/`<mac>`/ipv4/`<index>`/ip	|The IPv4 interface for the interface.	|v0|
+|/v0/meta-data/`<mac>`/ipv4/`<index>`/netmask	|The subnet mask for the IP configuration.	|v0|
+|/v0/meta-data/`<mac>`/ipv6/`<index>`/ip|The IPv6 address for the interface.	|v0
+|/v0/meta-data/`<mac>`/ipv6/`<index>`/netmask	|The subnet mask for the IP address.	|v0
+|/v0/meta-data/gateway|The instance gateway IP address.	|v0
+|/v0/meta-data/disks/`<index>`|Information about a single disk such as device name. For example, /dev/sda	|v0
+|/v0/meta-data/ssh-public-keys/`<number>`|SSH keys for the instance.|	v0
+|/v0/user-data	|User-data of the instance.	|v0
+|/v0/vendor-data	|Vendor-data for the instance.	|v0
+
+### Response Structure
+
+When a query to `/v0/meta-data` is dispatched, we return all fields newline-separated that are retrievable from that endpoint. Based on feedback from the Tinkerbell community, we might need to validate cloud-init behavior. Trailing slashes supposedly influence how the data is built in cloud-init.
+
+*Example*:
+```
+curl https://<example>:50061/v0/meta-data
+>
+hostname
+<mac>
+gateway
+disks
+ssh-public-keys
```

Comment:
The EC2 standard response includes `/` at the end of paths to indicate that they are traversable. 
This is done for array and map contents.

Reply:
Cloud-init's EC2 directory walking makes all this data available to jinja. I believe the path separators are used to hint the data type.

https://github.com/canonical/cloud-init/blob/main/cloudinit/sources/helpers/ec2.py#L74

Reply:
I thought we'd updated to include trailing slashes. That's the intent but I'll check and report if not.